tests/beta/tapes/test_qnode.py

```diff
@@ -0,0 +1,429 @@
+
+        def func(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return expval(qml.PauliZ(0))
+
+        qn = QNode(func, dev)
+
+        x = 0.12
+        y = 0.54
+
+        res = qn(x, y)
+
+        assert isinstance(qn.qtape, QuantumTape)
+        assert len(qn.qtape.operations) == 3
+        assert len(qn.qtape.observables) == 1
+        assert qn.qtape.num_params == 2
+
+        expected = qn.qtape.execute(dev)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # when called, a new quantum tape is constructed
+        old_tape = qn.qtape
+        res2 = qn(x, y)
+
+        assert np.allclose(res, res2, atol=tol, rtol=0)
+        assert qn.qtape is not old_tape
+
+    def test_jacobian(self, mocker, tol):
+        """Test the jacobian computation"""
+        spy = mocker.spy(QuantumTape, "numeric_pd")
+        dev = qml.device("default.qubit", wires=2)
+
+        def func(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return probs(wires=0), probs(wires=1)
+
+        qn = QNode(func, dev, h=1e-8, order=2)
+        assert qn.diff_options["h"] == 1e-8
+        assert qn.diff_options["order"] == 2
+
+        x = 0.12
+        y = 0.54
+
+        res = qn(x, y)
+        jac = qn.qtape.jacobian(dev, params=[0.45, 0.1])
+
+        assert jac.shape == (4, 2)
+
+    def test_returning_non_measurements(self):
+        """Test that an exception is raised if a non-measurement
+        is returned from the QNode."""
+        dev = qml.device("default.qubit", wires=2)
+
+        def func(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return 5
+
+        qn = QNode(func, dev)
+
+        with pytest.raises(
+            QuantumFunctionError, match="must return either a single measured observable"
+        ):
+            qn(5, 1)
+
+        def func(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return expval(qml.PauliZ(0)), 5
+
+        qn = QNode(func, dev)
+
+        with pytest.raises(
+            QuantumFunctionError, match="must return either a single measured observable"
+        ):
+            qn(5, 1)
+
+    def test_inconsistent_measurement_order(self):
+        """Test that an exception is raised if measurements are returned in an
+        order different to how they were queued on the tape"""
+        dev = qml.device("default.qubit", wires=2)
+
+        def func(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            qml.CNOT(wires=[0, 1])
+            m = expval(qml.PauliZ(0))
+            return expval(qml.PauliX(1)), m
+
+        qn = QNode(func, dev)
+
+        with pytest.raises(
+            QuantumFunctionError,
+            match="measurements must be returned in the order they are measured",

```

Comment:
The main reason behind this is that the tape does not care about the return order; it only cares about the order the measurements were queued. This is the order the numerical result will correspond to.

We could remove this check if we wanted, but then it could lead to unexpected behaviour for users, in that the QNode output will be ordered based on how the measurements queued, not the order they were returned in.

(Normally this is not an issue, since we almost always use the pattern `return expval(), var(),..`. This edge case is only possible if you queue, save to a variable, and then return like in the above example.)